packages/vite/src/node/server/persistentCache.ts

```diff
@@ -0,0 +1,388 @@
+        fileMap,
+        ssr
+      }
+
+      if (mod) {
+        entry.importedModules = Array.from(mod.importedModules)
+          .map((m) => m.url)
+          .filter(Boolean) as string[]
+        const importedBindings: any = {}
+        if (mod.importedBindings) {
+          for (const k in mod.importedBindings) {
+            const s = mod.importedBindings.get(k)
+            if (s) {
+              importedBindings[k] = Array.from(s)
+            }
+          }
+        }
+        entry.importedBindings = importedBindings
+
+        entry.acceptedHmrDeps = Array.from(mod.acceptedHmrDeps)
+          .map((m) => m.url)
+          .filter(Boolean) as string[]
+
+        entry.acceptedHmrExports = mod.acceptedHmrExports
+          ? (Array.from(mod.acceptedHmrExports).filter(Boolean) as string[])
+          : []
+
+        entry.isSelfAccepting = mod.isSelfAccepting
+      }
+
+      resolvedManifest.modules[key] = entry
+
+      queueManifestWrite()
+    } catch (e) {
+      logger.warn(
+        colors.yellow(
+          `Failed to write persistent cache entry '${key}' (${file}): ${e.message}`
+        )
+      )
+    }
+  }
+
+  return {
+    manifest: resolvedManifest,
+    getKey,
+    read,
+    write,
+    queueManifestWrite
+  }
+}
+
+interface ResolveServerPersistentCacheConfigPayload {
+  config: InlineConfig
+  pkgPath: string | undefined
+  resolvedRoot: string
+  resolvedConfigFile: string | undefined
+}
+
+export async function resolvePersistentCacheOptions(
+  payload: ResolveServerPersistentCacheConfigPayload
+): Promise<ResolvedServerPersistentCacheOptions | null> {
+  const { config, resolvedRoot, pkgPath, resolvedConfigFile } = payload
+
+  let resolvedServerPersistentCacheOptions: ResolvedServerPersistentCacheOptions | null
+  if (
+    config.experimental?.serverPersistentCaching != null &&
+    (config.experimental?.serverPersistentCaching !== false ||
+      (typeof config.experimental?.serverPersistentCaching === 'object' &&
+        config.experimental.serverPersistentCaching.enabled !== false))
+  ) {
+    const castedToObject =
+      typeof config.experimental?.serverPersistentCaching === 'object'
+        ? config.experimental.serverPersistentCaching
+        : null
+    const dir = castedToObject?.cacheDir
+      ? path.resolve(resolvedRoot, castedToObject.cacheDir)
+      : pkgPath
+      ? path.join(path.dirname(pkgPath), `node_modules/.vite-server-cache`)
+      : path.join(resolvedRoot, `.vite-server-cache`)
+
+    const cacheVersionFromFiles: string[] = (
+      castedToObject?.cacheVersionFromFiles ?? []
+    ).map((file) => path.join(resolvedRoot, file))
+
+    if (resolvedConfigFile) {
+      cacheVersionFromFiles.push(resolvedConfigFile)
+    }
+
+    const packageLockFile = lookupFile(
+      resolvedRoot,
+      [
+        'package-lock.json',
+        'yarn.lock',
+        'pnpm-lock.yaml',
+        'bun.lockb',
+        'npm-shrinkwrap.json'
+      ],
+      { pathOnly: true }
+    )
```

Comment:
Ideally this would be shared if the hash of `DepOptimizationMetadata`. But it's done a lot after the config resolution, so this would required some refactor to compute more cache related stuff early on.

Reply:
Maybe it's better to do that later when the feature is less experimental?

Reply:
Yes this was in part a comment for maintainers